chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+    }
+
+    /// Determines after how many key-values the process stops to
+    /// commit changes and to check interruptions.
+    const BATCH_SIZE: usize = 10_000;
+
+    let (parent_shard, status) = get_parent_shard_and_status(&resharder);
+    info!(target: "resharding", ?parent_shard, ?status, "flat storage shard split task: starting key-values copy");
+
+    // Parent shard flat storage head must be on block height just before the new shard layout kicks in.
+    // This guarantees that all deltas have been applied and thus the state of all key-values is up to date.
+    // TODO(trisfald): do this check, maybe call update_flat_storage_for_shard
+    let _parent_flat_head = status.flat_head;
+
+    // Prepare the store object for commits and the iterator over parent's flat storage.
+    let flat_store = resharder.runtime.store().flat_store();
+    let mut iter = flat_store.iter(parent_shard);
+
+    loop {
+        let mut store_update = flat_store.store_update();
+
+        // Process a `BATCH_SIZE` worth of key value pairs.
+        let mut iter_exhausted = false;
+        for _ in 0..BATCH_SIZE {
+            match iter.next() {
+                Some(Ok((key, value))) => {
+                    if let Err(err) =
+                        shard_split_handle_key_value(key, value, &mut store_update, &status)
+                    {
+                        error!(target: "resharding", ?err, "failed to handle flat storage key");
+                        return FlatStorageReshardingTaskStatus::Failed;
+                    }
+                }
+                Some(Err(err)) => {
+                    error!(target: "resharding", ?err, "failed to read flat storage value from parent shard");
+                    return FlatStorageReshardingTaskStatus::Failed;
+                }
+                None => {
+                    iter_exhausted = true;
+                }
+            }
+        }
+
+        // Make a pause to commit and check if the routine should stop.
+        if let Err(err) = store_update.commit() {
+            error!(target: "resharding", ?err, "failed to commit store update");
+            return FlatStorageReshardingTaskStatus::Failed;
+        }
+
+        // TODO(Trisfald): metrics and logs
+
+        // If `iter`` is exhausted we can exit after the store commit.
+        if iter_exhausted {
+            break;
+        }
+        if controller.is_interrupted() {
+            return FlatStorageReshardingTaskStatus::Interrupted;
+        }
+    }
+    FlatStorageReshardingTaskStatus::Successful
+}
+
+/// Handles the inheritance of a key-value pair from parent shard to children shards.
+fn shard_split_handle_key_value(
+    key: Vec<u8>,
+    value: FlatStateValue,
+    store_update: &mut FlatStoreUpdateAdapter,
+    status: &SplittingParentStatus,
+) -> Result<(), Error> {
+    if key.is_empty() {
+        panic!("flat storage key is empty!")
+    }
+    let key_column_prefix = key[0];
```

Comment:
I bet soon enough we'll have two bytes type indicators :) 
Nothing for you to do here, just being melodramatic ;)